### PR TITLE
Show HTML Help

### PR DESCRIPTION
**Pull request by rosen-vladimirov**
## Use case:
CLIs should be able to show HTML help when using ```<cli name> help <command name>```
When using ```<cli name> <command name> --help``` console help should be shown.
When an error is raised in command, only console help is shown.

## Implementation
* ```HelpCommand``` is modified to show console help when ```options.help``` is specified and html help when it is not.
* Add ```HtmlHelpService``` which handles both console and html help by parsing markdown help files.
* Each commands help is placed in a separate file. The files are markdown and they are used for both html and console help.
* Html pages are generated from markdown on post-install. As post-install command could fail before generating html pages, when html help should be shown, we check if command name is valid and if the command exists, but its html help is missing, we generate all html pages.
* Modify ```CommandsService``` to show console help by specifying options.help before calling ```HelpCommand```
* Modify ```DynamicHelpService``` and ```DynamicHelpProvider``` to accept isHtml option. When ```isHtml``` is specified, all localVariables will be set to true. This way html help will show full help, no matter of the conditions.
* Add ```basic_page.html``` page and used three placeholders in it. The page is used to generate all html files, by replacing the placeholders with commands name, help content and path to css file.
* Add ```styles.css``` file which contains styles used in html pages.
* Add definitions for marked module.
* Add ```appendFile``` method to ```FileSystem```
* Fix issue in yok, which was not requiring the first hierarchical command if it was not default one (if the hierarchical command doesn't have default command to execute) - this was leading to incorrect behavior when checking if command is valid.

http://teampulse.telerik.com/view#item/259417

### CODE REVIEW

File: services/html-help-service.ts [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+var chalk = require("chalk");
+
+export class HtmlHelpService implements IHtmlHelpService {
+	private static MARKDOWN_FILE_EXTENSION = ".md";
+	private static HTML_FILE_EXTENSION = ".html";
+	private static MAN_PAGE_NAME_PLACEHOLDER = "@MAN_PAGE_NAME@";
+	private static HTML_COMMAND_HELP_PLACEHOLDER = "@HTML_COMMAND_HELP@";
+	private static RELATIVE_PATH_TO_STYLES_CSS_PLACEHOLDER = "@RELATIVE_PATH_TO_STYLES_CSS@";
+
+	private pathToManPages: string;
```
Comment (Fatme) [inline services/html-help-service.ts]: Why do you need this local variable? You can use directly `staticConfig.HTML_PAGES_DIR`

Comment (rosen-vladimirov) [inline services/html-help-service.ts]: The reasons is caching of the result.
 staticConfig is executing ```path.join``` when getting HTML_PAGES_DIR and MAN_PAGES_DIR, I do not want to execute it every time.

File: services/html-help-service.ts [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+		return (() => {
+			var mdFileName = path.basename(pathToMdFile);
+			var htmlFileName = mdFileName.replace(HtmlHelpService.MARKDOWN_FILE_EXTENSION, HtmlHelpService.HTML_FILE_EXTENSION);
+			this.$logger.trace("Generating '%s' help topic.", htmlFileName);
+
+			var helpText = this.$fs.readText(pathToMdFile).wait();
+			var outputText = this.$microTemplateService.parseContent(helpText, { isHtml: true });
+			var htmlText = marked(outputText);
+
+			var filePath = pathToMdFile
```
Comment (Fatme) [inline services/html-help-service.ts]: `var filePath = path.join(this.pathToHtmlPages, htmlFileName` is shorter.

Comment (rosen-vladimirov) [inline services/html-help-service.ts]: yep, I'm not sure what I've been thinking here :smile_cat:

File: services/html-help-service.ts [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+				.replace(path.basename(this.pathToManPages), path.basename(this.pathToHtmlPages))
+				.replace(mdFileName, htmlFileName);
+			this.$logger.trace("HTML file path for '%s' man page is: '%s'.", mdFileName, filePath);
+
+			var outputHtml = basicHtmlPage
+				.replace(HtmlHelpService.MAN_PAGE_NAME_PLACEHOLDER, mdFileName.replace(".md", ""))
```
Comment (Fatme) [inline services/html-help-service.ts]: You can use `MARKDOWN_FILE_EXTENSION` constant here.

File: services/html-help-service.ts [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+	private readManFileForCommand(commandName: string): IFuture<string> {
+		return ((): string => {
+			var mdFileName = this.convertCommandNameToFileName(commandName) + HtmlHelpService.MARKDOWN_FILE_EXTENSION;
+			this.$logger.trace("Reading help for command '%s'. FileName is '%s'.", commandName, mdFileName);
+			var helpText: string;
+
+			var markdownFile = _.find(this.$fs.enumerateFilesInDirectorySync(this.pathToManPages), file => path.basename(file) === mdFileName);
+			if(markdownFile) {
+				helpText = this.$fs.readText(markdownFile).wait();
+			} else {
```
Comment (Fatme) [inline services/html-help-service.ts]: `else` is unneeded. Just `if(markdownFile) {
				return this.$fs.readText(markdownFile).wait(); 
}
this.$errors.failWithoutHelp("Unknown command '%s'. Try '$ %s help' for a full list of supported commands.", mdFileName.replace(".md", ""), this.$staticConfig.CLIENT_NAME.toLowerCase());
`

File: services/html-help-service.ts (hunk 2 above)
Comment (rosen-vladimirov) [inline services/html-help-service.ts]: Ahh, I've got it - as I want to preserve same directory structure, I need the replace method. If I use your suggestion, all html files will be placed inside html directory.

File: services/html-help-service.ts [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+			this.$logger.trace("Found page to open: '%s'", pageToOpen);
+			this.$opener.open(pageToOpen);
+			return true;
+		} 
+		
+		this.$logger.trace("Unable to find file: '%s'", htmlPage);
+		return false;
+	}
+
+	private readManFileForCommand(commandName: string): IFuture<string> {
```
Comment (teobugslayer) [inline services/html-help-service.ts]: Maybe the better name is "readMdFileForCommand", thoughts?

**Comment by teobugslayer** [comment]
Cool, :+1:

**Pull request merged**